geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java

```diff
@@ -644,27 +661,104 @@ private static String urlDecUtf8(String value) {
      *     java.lang.String)
      */
     public void putLayerMetadata(final String layerName, final String key, final String value) {
-        Properties metadata = getLayerMetadata(layerName);
+        // Get current metadataFile
+        final File metadataFile = getMetadataFile(layerName);
+        // Get last modification time for metadata file
+        long lastModDate = metadataFile.lastModified();
+        // Get metadata content from current metadataFile
+        Properties metadata = getLayerMetadata(metadataFile);
+
+        // Evaluate updating key arg related values (remove/add/update or skip)
+        boolean skipUpdate = false;
+        // Note: since we want to avoid writing metadata file if nothing changed, then we need to
+        // check if it's required
+        // or not (flag skipUpdate). That is reason not to use metadata.merge method instead of
+        // using setProperty
         if (null == value) {
             metadata.remove(key);
         } else {
+            String currValue = metadata.getProperty(key);
+            String encodedValue = null;
             try {
-                metadata.setProperty(key, URLEncoder.encode(value, "UTF-8"));
+                encodedValue = URLEncoder.encode(value, "UTF-8");
             } catch (UnsupportedEncodingException e) {
                 throw new RuntimeException(e);
             }
+            skipUpdate =
+                    currValue != null && encodedValue != null && currValue.equals(encodedValue);
+            // If existing value for key is the same as value arg, then avoid to write it
+            if (!skipUpdate) {
+                metadata.setProperty(key, encodedValue);
+            } else {
+                skipUpdate = true;
+            }
         }
 
-        final File metadataFile = getMetadataFile(layerName);
+        // We call to save metadata only if has changed
+        if (!skipUpdate) {
+            // Write metadata in a temp file
+            File tempFile = null;
```

Comment:
This entire block could be in its own method, "writeOptimisticLock", the name would also give away what it is intending to do.